storage/queue.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package storage provides implementations and shared components for tessera storage backends.
+package storage
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/globocom/go-buffer"
+)
+
+// Queue knows how to queue up a number of entries in order, taking care of deduplication as they're added.
+//
+// When the buffered queue grows past a defined size, or the age of the oldest entry in the
+// queue reaches a defined threshold, the queue will call a provided FlushFunc with
+// a slice containing all queued entries in the same order as they were added.
+//
+// If multiple identical entries are added to the queue between flushes, the queue will deduplicate them by
+// passing only the first through to the FlushFunc, and returning the index assigned to that entry to all
+// duplicate add calls.
+// Note that this deduplication only applies to "in-flight" entries currently in the queue; entries added
+// after a flush will not be deduped against those added before the flush.
+type Queue struct {
```

Comment:
I think of those as "sequenced entries", there's a process migrates them to a different storage and integrates them into the tree but I'm not really thinking of that as a queue.

I don't want to call this struct `Batch` since that's not what it is, I thought about using `Pool` or `Buffer` but those names lack the connotation of FIFO which `Queue` carries.